### ISSUE TITLE
Extra arguments passed to non-variadic command are swallowed

### ISSUE THREAD

**Issue by mojavelinux** [tj/commander.js#749]
If an action-based command does not have a variadic argument, extra arguments supplied by the user are swallowed. I would expect either for the command to exit with an error, or for those extra arguments to be passed to the callback method so they can be handled appropriately.

The root of the problem is that the program is overwriting the extra entry in the args array.

**Comment by mojavelinux** [comment]
Here's a simple example:

```
cli
  .name('cli')
  .command('action [value]')
  .action((value, cmd) => {
    console.log('action: ' + cmd.name())
    console.log('value: ' + value)
  })

cli.parse(process.argv)
```

Call this with:

```
cli action value extra
```

No warning. And the value "extra" is nowhere to be seen in the model except in rawArgs.

**Comment by mojavelinux** [comment]
The open question is whether we should consider these extra arguments an error condition or we should simply pass them to the action to handle. I've implemented it as an extra argument, though the code could easily repurposed.

**Comment by twifty** [comment]
Here's another example:

```js
cli
    .arguments('<command> [args...]')
    .option('-h, --host <host>', 'The host on which the server should listen')
    .option('-p, --port <port>', 'The port number on which the server should listen')
    .action((command) => {
        runCommand(command, args, cli.host, cli.port)
    })
    .parse(process.argv);
```

Called with `cli php -v`.

The '-v' is swallowed. I worked around the issue by looking for command in the rawArgs array then slicing from the index.

**Comment by shadowspawn** [comment]
I was chasing some different issues, and independently came to same conclusion as @mojavelinux:

> The root of the problem is that the program is overwriting the extra entry in the args array.

Working on this currently.

**Comment by shadowspawn** [comment]
There are improvements on the develop branch for the next release (#1030 and #1048).

**Comment by shadowspawn** [comment]
v4.0.0-0 prerelease published: #1067

**Comment by shadowspawn** [comment]
v4.0.0 has been released with a fix for program.args which was getting the command appended (matching the arguments passed to the action handler).
